lib/Dialect/FIRRTL/FIRRTLDialect.cpp

```diff
@@ -84,6 +85,55 @@ Operation *FIRRTLDialect::materializeConstant(OpBuilder &builder,
   return nullptr;
 }
 
+namespace {
+// During canonicalization, constant folding can result in operand types
+// changing from non-'const' to 'const'. This can cause result types to likewise
+// change their inference from non-'const' to 'const'. This pattern handles
+// those result type changes.
+struct ReinferResultTypes
+    : public mlir::OpInterfaceRewritePattern<mlir::InferTypeOpInterface> {
+  ReinferResultTypes(MLIRContext *context)
```

Comment:
This seems a little heavy-weight to be run on every op every time it changes?

IIRC we have a cheaper version of this that avoids the need for the vector/etc, which might be cheaper to use for those ops.  Don't know it has an interface though, so maybe nevermind.

Reply:
This was the best option I could find that is compatible with how the canonicalizer works. I'm definitely open to other ideas.

Reply:
The issue is that mutating an operation during a fold is only allowed to affect that operation. But if an operation's result type has changed to const, downstream consumers of that const value might now infer a const result as well.